go-controller/pkg/libovsdbops/template_var.go

```diff
@@ -0,0 +1,231 @@
+package libovsdbops
+
+import (
+	"context"
+	"regexp"
+	"strings"
+
+	libovsdbclient "github.com/ovn-org/libovsdb/client"
+	libovsdb "github.com/ovn-org/libovsdb/ovsdb"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+)
+
+const TemplatePrefix = "^"
+
+// Chassis_Template_Var records store (for efficiency reasons) a 'variables'
+// map for each chassis. For simpler CMS code, store templates in memory
+// as a (Name, Value) tuple where Value is a map with potentially different
+// values on each chassis.
+type Template struct {
+	Name string
+
+	// Per chassis template value, indexed by chassisID.
+	Value map[string]string
+}
+
+type TemplateMap map[string]*Template
+type ChassisTemplateVarMap map[string]*nbdb.ChassisTemplateVar
+
+// Len returns the number of chasis on which this Template variable is
+// instantiated (has a value).
+func (t *Template) Len() int {
+	return len(t.Value)
+}
+
+// ToReferenceString returns the textual representation of a template
+// reference, that is, '^<template-name>'.
+func (t *Template) ToReferenceString() string {
+	return TemplatePrefix + t.Name
+}
+
+// IsTemplateReference returns true if 'name' is a valid template reference.
+func IsTemplateReference(name string) bool {
+	return strings.HasPrefix(name, TemplatePrefix)
+}
+
+// TemplateNameFromReference extracts the template name from a textual
+// reference to a template.
+func TemplateNameFromReference(template string) string {
+	return strings.TrimPrefix(template, TemplatePrefix)
+}
+
+// MakeTemplateName creates a valid template name by replacing invalid
+// characters in the original 'name' with '_'. Existing '_' are doubled.
+func MakeTemplateName(name string) string {
+	invalidChars := regexp.MustCompile(`[/\-$@]`)
+	name = strings.Replace(name, "_", "__", -1)
+	return invalidChars.ReplaceAllString(name, "_")
+}
+
+// MakeTemplate intializes a named Template struct with 0 values.
+func MakeTemplate(name string) *Template {
+	return &Template{Name: name, Value: map[string]string{}}
+}
+
+func forEachNBTemplateInMaps(templateMaps []TemplateMap, callback func(nbTemplate *nbdb.ChassisTemplateVar)) {
+	// First flatten the maps into *nbdb.ChassisTemplateVar records:
+	flattened := ChassisTemplateVarMap{}
+	for _, templateMap := range templateMaps {
+		for _, template := range templateMap {
+			for chassisName, templateValue := range template.Value {
+				if templateValue == "" {
+					continue
+				}
+				nbTemplate, found := flattened[chassisName]
+				if !found {
+					nbTemplate = &nbdb.ChassisTemplateVar{
+						Chassis:   chassisName,
+						Variables: map[string]string{},
+					}
+					flattened[chassisName] = nbTemplate
+				}
+				nbTemplate.Variables[template.Name] = templateValue
```

Comment:
I think I prefer the current one, to avoid a second lookup (`flattened[chassisName]`), exactly because `nbTemplate` is a pointer.  I'll leave it as is if that's OK with you.